packages/flutter/lib/src/painting/text_painter.dart

```diff
@@ -235,15 +236,67 @@ class TextPainter {
   double get preferredLineHeight {
     if (_layoutTemplate == null) {
       final ui.ParagraphBuilder builder = ui.ParagraphBuilder(
-        _createParagraphStyle(TextDirection.rtl),
-      ); // direction doesn't matter, text is just a space
+        _createParagraphStyle(TextDirection.ltr),
+      ); // direction doesn't matter, text is just a space, ltr is more common.
       if (text?.style != null)
         builder.pushStyle(text.style.getTextStyle(textScaleFactor: textScaleFactor));
       builder.addText(' ');
       _layoutTemplate = builder.build()
         ..layout(ui.ParagraphConstraints(width: double.infinity));
     }
-    return _layoutTemplate.height;
+    return max(_layoutTemplate.height, _maxLineHeight);
+  }
+
+  // Tracks the max line height to prevent jitter in text box size.
+  //
+  // When a taller-than-existing glyph is typed, prevent the height from
+  // shrinking smaller than it.
+  double _maxLineHeight = 0;
+
+  /// Attempt to get the actual height of the line after layout.
+  ///
+  /// This returns the max height of the line that the character at [offset]
+  /// belongs in. When the paragraph has not been laid out, this returns the
+  /// estimate given by [preferredLineHeight] as a fallback.
+  ///
+  /// The [offset] parameter is the byte index of the text the height should
+  /// be queried at.
+  ///
+  /// This method may produce different heights than [preferredLineHeight] as
+  /// this is the actual measured height of the laid out text. This means that
+  /// the height of each line can vary depending on the maximum heights of
+  /// the characters in the line.
+  double preferredLineHeightAtOffset(int offset) {
+    if (_needsLayout) {
+      return preferredLineHeight;
+    }
+    final double height = _getHeightFromUpstream(offset)
+        ?? _getHeightFromDownstream(offset)
+        ?? preferredLineHeight;
+    _maxLineHeight = max(height, _maxLineHeight);
+    return height;
+  }
+
+  double _getHeightFromUpstream(int offset) {
+    final int prevCodeUnit = _text.codeUnitAt(offset - 1);
+    if (prevCodeUnit == null)
+      return null;
+    final int prevRuneOffset = _isUtf16Surrogate(prevCodeUnit) ? offset - 2 : offset - 1;
+    final List<TextBox> boxes = _paragraph.getBoxesForRange(prevRuneOffset, offset, boxHeightStyle: ui.BoxHeightStyle.max);
+    if (boxes.isEmpty)
+      return null;
+    return boxes.last.bottom - boxes.last.top;
+  }
+
+  double _getHeightFromDownstream(int offset) {
+    final int nextCodeUnit = _text.codeUnitAt(offset + 1);
+    if (nextCodeUnit == null)
+      return null;
+    final int nextRuneOffset = _isUtf16Surrogate(nextCodeUnit) ? offset + 2 : offset + 1;
```

Comment:
what if it's a high surrogate? wouldn't that mean you would have to only add 1? i'm confused as to what this is trying to do.